hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/RSRpcServices.java

```diff
@@ -308,18 +310,35 @@
    */
   private static final long DEFAULT_REGION_SERVER_RPC_MINIMUM_SCAN_TIME_LIMIT_DELTA = 10;
 
-  /*
+  /**
    * Whether to reject rows with size > threshold defined by
    * {@link RSRpcServices#BATCH_ROWS_THRESHOLD_NAME}
    */
   private static final String REJECT_BATCH_ROWS_OVER_THRESHOLD =
     "hbase.rpc.rows.size.threshold.reject";
 
-  /*
+  /**
    * Default value of config {@link RSRpcServices#REJECT_BATCH_ROWS_OVER_THRESHOLD}
    */
   private static final boolean DEFAULT_REJECT_BATCH_ROWS_OVER_THRESHOLD = false;
 
+  /**
+   * Determine the bootstrap nodes we want to return to the client connection registry.
+   * <ul>
+   * <li>{@link #MASTER}: return masters as bootstrap nodes.</li>
```

Comment:
Why do we even need this? I thought the agreement was to not let master be in the hot path? For backward compatibility sakes?

Reply:
I think at least for a small cluster, using both masters and region servers as bootstrap node is a valid requirements. And I'm not sure if someone wants to use only masters, but at least you guys used to make use of backup masters and it works fine, I think it is OK to provide a MASTER option here, not a big deal, the default value is REGIONSERVER and usually users will not change it.

Reply:
Sorry for the delay boss, I was busy last few days.

My only concern here is that it is against the agreed design principle of not putting masters inline with the connection path and exposing such a configuration might complicate things going forward as future developers might build on top of it. 

Typically for small clusters, since it is impossible to run a meaningful cluster without a region server, but things can run for sometime without a master, so we can expect that at least one region server to up all the time if it is serving any load. So a registry can be served with that region server(s)? WDYT?

@saintstack / @francisliu you have any comments on this approach?

Reply:
I'm trying to figure out what the approach is. It seems to be moving Registry host to RegionServers -- which sounds good but I'm not sure how we get off the ground... bootstrap. And yeah, would be good to avoid Masters if we can.

Reply:
Looking more, appreciate offering the option but like the @bharathv  comment that we should just do RS.

Reply:
I'm fine with removing this but then we can not deprecated MasterRegistry? Or we just tell users you can not use masters as registry any more? Since the MasterRegistry has already been released out, and people are currently using it, I'm not sure whether this is a good idea...
For me I prefer we set the default value to only use region servers, but if users do want to use masters, they could change the config as they like, especially that the implementation itself does not introduce much complexity.
So let's make a decision here? Let's move forward.
Thanks.

Reply:
> but if users do want to use masters

I don't see how it breaks current semantics of MasterRegistry if they continue using it.  If some users are currently using Master Registry, the code flow and request path still remains the same. Following is the code in MasterRegistry (based on your last commit)

```

  @Override
  protected Set<ServerName> getBootstrapNodes(Configuration conf) throws IOException {
    return parseMasterAddrs(conf);
  }

  @Override
  protected CompletableFuture<Set<ServerName>> fetchEndpoints() {
    return getMasters();
  }
```

So the flow is, initial bootrap list = whatever user configured and using that bootstrap list we call fetchEndPoints() which return a bunch fo master stubs (_only masters_, nothing else). Now MasterRpcServers implements the RegistryProtos.ClientMetaService (via RsRpcServices) so we have all the methods available and the request path is the same? 

With this current change, (my understanding) RpcConnectionRegistry behavior actually changes, there we have 

```
private CompletableFuture<Set<ServerName>> getBootstrapNodes() {
    return this
      .<GetBootstrapNodesResponse> call(
        (c, s, d) -> s.getBootstrapNodes(c, GetBootstrapNodesRequest.getDefaultInstance(), d),
        r -> r.getServerNameCount() != 0, "getBootstrapNodes()")
      .thenApply(RpcConnectionRegistry::transformServerNames);
  }

  @Override
  protected CompletableFuture<Set<ServerName>> fetchEndpoints() {
    return getBootstrapNodes();
  }
```

The above code can potentially return MASTERS if it is configured on the server side. Here we are _not_ breaking anything because `RpcConnectionRegistry` is a new addition and the semantics are to use regionservers. As I mentioned above, MasterRegistry still uses getMasters(), so it never sees regionservers. I don't see how anything is broken.

If someone wants to use masters, they can just use MasterRegistry until it is deprecated in 4.0.0.

Reply:
What I propose is deprecated MasterRegistry and remove it in the future, so we'd better provide the same ability in RpcConnectionRegistry. If you want to keep MasterRegistry there, then things will be different. Maybe we should rename RpcConnectionRegistry to RegionServerRegistry, and all the configuration names should be changed. What is your expection here? I'm fine with both, just let me know...

Reply:
And the plan here is to deprecate MasterRegistry in 2.5.0, not 4.0.0, we will remove it in 4.0.0...

Reply:
Ah ok. It appeared that you intended to keep both MasterRegistry, RpcConnectionRegistry with masters. Misunderstanding.

> Maybe we should rename RpcConnectionRegistry to RegionServerRegistry, and all the configuration names should be changed.

To me this sounds like something we should do, explicitly communicates that we want RS to be the bootstrap points and master is no where. So +1 on this. Not sure what others think or if they have a different preference. To me its ok to not change the configuration names, but I'm also if you plan to change them.


> And the plan here is to deprecate MasterRegistry in 2.5.0, not 4.0.0, we will remove it in 4.0.0...

Ya thats what I meant, typo

Reply:
> It appeared that you intended to keep both MasterRegistry, RpcConnectionRegistry with masters.

Let me explain more...

The issue here, aims to remove MasterRegistry in the future, so
1. Ideally RpcConnectionRegistry should cover what we have in MasterRegistry, by some configurations at least.
2. We just do not want to support masters any more, so we do not need to provide configurations at server side, after MasterRegistry is removed, users can not use masters as connection registry endpoint any more.

But if we go with option 2, maybe it breaks some users, so maybe we should keep MasterRegistry and do not mark it as deprecated? And then, the RpcConnectionRegistry should better be renamed to RegionServerRegistry.

So let me conclude, there are 3 options:
1. Deprecated MasterRegistry, plan a removal in 4.0.0, and let RpcConnectionRegistry still provide the ability to use masters as connection registry endpoint.
2. Deprecated MasterRegistry, plan a removal in 4.0.0, but still keep RpcConnectionRegistry only use region servers as connection registry endpoint.
3. Do not deprecated MasterRegistry, rename RpcConnectionRegistry to RegionServerRegistry.

I do not think we need to rename RpcConnectionRegistry to RegionServerRegistry if we still want to remove MasterRegistry, so there is option 4.

Please let me know your choice. But if you want to choose 2 or 3, I think we'd better post a discussion email to the dev list, as it removes a feature currently in use, we need to collect more information from users.

Thanks.

Reply:
What I'm saying in my above comment is "deprecate MasterRegistry and rename RpcConnectionRegistry" (similar to option 2) and _everything still works as expected for current MasterRegistry users_. Details in https://github.com/apache/hbase/pull/3566#discussion_r687025151

If we go with option 2, can you explain me what does not work for current users of MasterRegistry? If you explain how it is broken, I'm +1 on option (1). (see my last comment as to why I think it is not broken, may be I misunderstood something) https://github.com/apache/hbase/pull/3566#discussion_r687025151

Only broken case I can think of is old client binary is not compatible with new Server binary, but that is not a problem we are worried about? Is there any other case too?
Thanks.

Reply:
There are no compatible issue between client and server...

The key difference between option 1 and 2, is that whether still support using master as registry endpoint after the removal of MasterRegistry in 4.0.0. If we still support this feature in the future, then I think a release note and a notice email to dev list is enough. If we really think we should remove this feature, then we'd better send a discussion email to dev list first. If there are no big concerns, then we can do this. If no, I think we'd better go with option 1, to let RpcConnectionRegistry still has the ability to use masters as registry endpoint.

Is this clear enough? Let me point out the key problem again, it is about whether people could still use masters as registry endpoint in 4.0.0, not about now...

Thanks.

Reply:
ya okay. I understand your concern now. Thanks for the explanation.

>  If we really think we should remove this feature, then we'd better send a discussion email to dev list first. If there are no big concerns, then we can do this.

+1 on this fwiw. Since the decision to inline masters in RW path was a wrong design choice, I vote to get rid of it in the long term.

Patch lgtm otherwise.

Reply:
Email sent.